### PR TITLE
isde-project-2023-C development is completed

### LINKED ISSUE #1: Image Histogram
From isde-projects-2023-BASE created by [maurapintor](https://github.com/maurapintor): unica-isde/isde-projects-2023-BASE#1  

## Summary of the request

Create a new handle that allows picking one image from a list and returns the histogram of an image. The histogram can be rendered either in the frontend (JavaScript library, for extra points), or in the backend and passed as an image (python / matplotlib).
**This is a new feature. Old functionalities must be preserved.**
Difficulty: 
image/image(If using Javascript)

## Subtasks
* Copy the classification API form and modify it where it is required.
* Implement the new API that takes as input the id of an image and returns the histogram.
* Visualize the histogram and the image in the GUI. 

## Plan for the implementation

Modify the code such that the new interface shows the dropdown menu to select the image and the submit button.
It is not required to use queues. The API can be implemented as showing the form if a GET request is issued, and the results if the POST is validated.

image

Useful links: 
* https://stackoverflow.com/questions/22159160/python-calculate-histogram-of-image

### LINKED ISSUE #2: Image Transformation
From isde-projects-2023-BASE created by [maurapintor](https://github.com/maurapintor): unica-isde/isde-projects-2023-BASE#2  

## Summary of the request

Create a handler that allows picking an image and set parameters for an image transformation (https://pillow.readthedocs.io/en/stable/reference/ImageEnhance.html). The user should be allowed to specify the color, brightness, contrast, and sharpness to apply to the original image. The transformed image should appear in the output interface.
**This is a new feature. Old functionalities must be preserved.**
Difficulty:
image

## Subtasks

* Copy the classification API interface.
* Create a form for selecting the transformation.
* Implement the API that applies the transformation to the selected image.
* Implement the visualization of results.

## Plan for the implementation

Modify the code such that the new interface shows the dropdown menu to select the image, the submit button, and a NEW field that allows specifying values for the transformation to apply. The values for the transformation should have default values, but should be editable by the user. Use forms to create the interface for selecting the transformation. Use correct field types for the inputs.
It is not required to use queues. The API can be implemented as showing the form if a GET request is issued, and the results if the POST is validated.

Useful links:
* https://pillow.readthedocs.io/en/stable/reference/ImageEnhance.html
* https://fastapi.tiangolo.com/tutorial/request-forms/

### LINKED ISSUE #3: Download results button
From isde-projects-2023-BASE created by [maurapintor](https://github.com/maurapintor): unica-isde/isde-projects-2023-BASE#3  

## Summary of the request

Add the “download results” button to the web interface. The button should allow the user to download the results (classification scores) as a JSON file and as a png file showing the top 5 scores in a plot (bar chart).
**All the functionalities must be preserved, this is just a new feature**.
Difficulty:
image

## Subtasks

* Add the "download results" button to the interface.
* Add the "download plot" button to the interface.
* Implement the API for returning a JSON file with the results.
* Implement the API for returning the plot.
* Connect the buttons with the APIs.

## Plan for the implementation

The easiest way to implement the new functionality is to create the files on the go when the button is pressed, through the button URL linked with an HTML tag (```<a href="URL">```, see references below).

image

Useful links:
* https://www.w3schools.com/tags/tag_a.asp
* https://stackoverflow.com/questions/60716529/download-file-using-fastapi

### LINKED ISSUE #4: Upload image button
From isde-projects-2023-BASE created by [maurapintor](https://github.com/maurapintor): unica-isde/isde-projects-2023-BASE#4  

## Summary of the request

Add the “upload image” functionality. Implement a new route that instead of showing the dropdown list of images to select, allows the user to select an image from the computer, and returns the classification results.
The button should allow the user to upload a custom image and classify it. The image should be displayed in the output interface along with the classification scores.
**The old functionalities should be preserved, this is an additional feature**.
Difficulty:
image

## Subtasks

* Copy the route and HTML file of the classification API.
* Add the new route to the routes init file.
* Add "upload" button in the frontend (HTML).
* Implement the API for storing the image.
* Pass the image to the classification API.

## Plan for the implementation

image

Implement using the FastAPI functionalities and modifying a copy of the available classification API. No particular additional library is required.

Useful links:
* https://fastapi.tiangolo.com/tutorial/request-files/

### PR DESCRIPTION
**Pull request by CipherB7R**
All the requested features have been implemented successfully by the isde-project-2023-C team.

You can see the tracked progress inside our Github's forked repo page.

Each requirement was assigned to one team member but we worked in pairs to solve the bugs that appeared in our code: 
some of the most important ones are documented in this changelog, together with their solutions; we leave them here, as we believe the students of the next-to-come editions of this course may benefit from it.

Changelog:

Added the following features:

> Image transformation  (Resolves #2)
> Image histogram (Resolves #1)
> Upload image button (Resolves #4)
> Download results button (Resolves #3)

Changed the following files:

> utils.py  >>> Added an asyncronous image save function  
> base.html >>> Added several navbar buttons to access the new features
> home.html >>> Added a link to the histogram generation page
> classification_output.html >>> Added the functionalities relative to the download results feature
> main.py   >>> Added the python functions to handle the HTTP requests relative to the new features
> requirements.txt >>> Added matplotlib to the libraries required for the project

Added the following files:

> transform_image.html 		>>> page to select the image to be transformed
> transform_image_output.html	>>> page that shows the transformed image (image is transformed server-side)
> histogram_select.html		>>> page to select the image for which a histogram shall be generated 
> histogram_output.html	>>> page that shows the generated histogram (histogram is generated client-side)
> graph_histogram.js		>>> contains the javascript code for the client-side histogram generation
> classification_upload.html	>>> page that enables an user to upload an JPEG image

> forms folder 	>>> contains the model classes for handling and sanitizing user input coming from the requested
                                         features' most complex forms
>> classification_form_upload.py    >>> Handles the sanitization of the upload form and allows to save the uploaded image
>> transform_image_form.py 	 >>> Handles the sanitization of the transform form and allows (through the
                                                                  transform_wrapper class) the transformation of the selected image

> image_transform folder  >>> contains the "core" python classes for the image transformation utility
>> transform_wrapper.py  >>> contains the code which handles the transformation of an image

Other:

> Added FastAPI documentation to all the new features' functions inside main.py file

Suggestions:

> In order to not overwrite a previously uploaded image, we need an efficient way to check if that
  image already exists: this can be done by hashing the image files when they get uploaded to the
  server and saving all the hashes in a separate csv file. This suggestion wasn't implemented
  since modifying the prepare_models.py file would violate the requirements of the project 
  (namely, the "don't change files that don't need to be changed to implement the functionalities" part).

Notable bugs encountered and how they were solved:

> The histogram of the image is generated locally by the client with javascript. 
  With big and high definition images, the function which generated the histogram would fire before
  the image could be downloaded from the server. This resulted in a blank histogram.

  We solved this bug by making the function fire only after the image gets fully downloaded.
  This is done by attaching the function to the (on)load event of the image.

> The user could not transform the same image and get the different versions of it.
  Apparently, the browser would just load the first version from the cache, even if the user changed the transformation parameters.
  We found out that just setting the cache-control policy of the HTTP response would not work on some browsers.
  
  We solved this by forcing the transformed image download: namely, sending it to the client as base64 encoded
  data embedded in the response HTML.

### CODE REVIEW

**Review by asotgiu — CHANGES_REQUESTED** [review]
I reviewed the code and tested the app: the code is well documented and, except for the following issues, the requested functionalities are correctly implemented. 
Please fix them and request a new review when it's done. There's no need to open a new PR, it's enough to push the new commits to the origin repository (yours) of the PR. In that way, I will see only the new changes and I won't have to review the full code again.

#### Image uploading
The current implementation of the image uploading raises some critical issues (note that persistently saving the uploaded images in memory and making them available among the pre-loaded images was not required, but it's ok to do so):

1. if a user uploads a file with the same name as a pre-loaded image, it will be overwritten. This does not preserve the original functionality of the app, the pre-loaded images should not be overwritten or modified;
2. if multiple users upload files with the same name, the app may behave unpredictably;
3. if a user uploads a file that is not an image, but has the `.JPEG` extension, the app crashes and the file remains in memory anyway. The error should be handled (for instance with a try-except) and the file should be removed.
Additional suggestion (you don't have to implement it): it is not safe to check the file format by only looking at its extension. A safer method is to check the magic number, for instance with [python-magic](https://github.com/ahupp/python-magic).

#### Histogram plot
The histogram is not always shown, at the first request, for a large number of images. I had this issue with Chrome and Brave browsers, whereas it never happened with Firefox.

Please, link the resolved issues to the pull request changelog (edit the description of the PR, or at least use the GUI functionality to link the PR to the issues of the target repository).
Thanks!

**Comment by CipherB7R** [comment]
The assignees of the mentioned faulty functionalities corrected the relative issues.

Regarding Image Uploading:

noting that "persistently saving the uploaded images in memory and making them available among the pre-loaded images was not required", all saving operations were removed from the codebase.

The app now keeps them in working memory just for the time needed to classify them and build the output page.
This new implementation should automatically deal with points 1 and 2 (since the images are retained in working memory in separate objects by means of the underlying SpooledTemporaryFile implementation, see next). 
Point 3 is dealt by means of the underlying implementation of the UploadFile class (a SpooledTemporaryFile):
the UploadFile is closed (and its underlying temporary file deleted) whenever computation terminates, an exception is thrown or by the garbage collector itself (in case of unexpected exceptions).

The app now checks for the uploaded file's magic numbers to recognize the content of the file; in case this control fails, we handle the exception regarding the creation of the Image object to be classified, by throwing the user back to the upload page with an error.


Regarding Histogram plot:

The issue was related to the onload event and how those different browsers manage it (Chrome seems to let it fire only once, for the document). In this last version it seems to work with all browsers.

File: app/forms/classification_form_upload.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        form = await self.request.form()
+        self.image_file = form.get("immagine")
+        self.model_id = form.get("model_id")
+        self.image_id = self.image_file.filename
+
+    async def is_valid(self):
+        # FastAPI bug https://github.com/tiangolo/fastapi/discussions/9705
+        if not self.image_file or not isinstance(self.image_file, starlette.datastructures.UploadFile) \
+                or not self.image_file.filename.endswith(".JPEG"):  # utils.list_images() accepts only '.JPEG' extension
+            self.errors.append("A valid .JPEG image is required (check file extension, it must be uppercase too!)")
```
Comment (asotgiu) [inline app/forms/classification_form_upload.py:33]: Is this (strong) constraint still required?

**Review by asotgiu — COMMENTED** [review]
All the issues are fixed, well done!
I confirm the histogram now is always loaded in my browsers.
I added a small comment, if you want to check it would be great, otherwise I will merge the PR as is.
Thank you!

File: app/forms/classification_form_upload.py (hunk 1 above)
Comment (CipherB7R) [reply app/forms/classification_form_upload.py:33]: That, plus the reason mentioned in the comment next to the lines of code.

**Review by asotgiu — APPROVED** [review]

**Pull request merged**